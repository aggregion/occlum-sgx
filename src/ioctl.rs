use std::{ffi::CString, sync::Mutex};

use log::trace;

use crate::{
    constants::{
        IOCTL_GEN_DCAP_QUOTE, IOCTL_GET_DCAP_QUOTE_SIZE, IOCTL_GET_DCAP_SUPPLEMENTAL_SIZE,
        IOCTL_VER_DCAP_QUOTE,
    },
    error::SGXError,
    types::{SGXQuoteVerifyResult, SGXReportData},
};

#[repr(C)]
#[derive(Debug)]
pub struct IoctlGenDCAPQuoteArg {
    pub report_data: *const SGXReportData, // Input
    pub quote_size: *mut u32,              // Input/output
    pub quote_buf: *mut u8,                // Output
}

#[repr(C)]
pub struct IoctlVerDCAPQuoteArg {
    pub quote_buf: *const u8,                                 // Input
    pub quote_size: u32,                                      // Input
    pub collateral_expiration_status: *mut u32,               // Output
    pub quote_verification_result: *mut SGXQuoteVerifyResult, // Output
    pub supplemental_data_size: u32,                          // Input (optional)
    pub supplemental_data: *mut u8,                           // Output (optional)
}

lazy_static! {
    pub static ref IOCTL_CLIENT: Mutex<IoctlClient> = {
        let client = IoctlClient::new();
        Mutex::new(client)
    };
}

// Client which send ioctls to the Occlum LibOS
// @see https://github.com/occlum/occlum/blob/master/src/libos/src/fs/dev_fs/dev_sgx/mod.rs
pub struct IoctlClient {
    fd: Option<i32>,
    quote_size: Option<u32>,
    supplemental_size: Option<u32>,
}

impl IoctlClient {
    fn new() -> Self {
        Self {
            fd: None,
            quote_size: None,
            supplemental_size: None,
        }
    }

    fn fd(&mut self) -> Result<i32, SGXError> {
        if self.fd.is_none() {
            let path = CString::new("/dev/sgx").expect("CString::new failed");
            let fd = unsafe { libc::open(path.as_ptr(), libc::O_RDONLY) };

            if fd <= 0 {
                return Err(SGXError::DeviceOpenFailed("/dev/sgx"));
            }

            self.fd = Some(fd);
            Ok(fd)
        } else {
            Ok(self.fd.as_ref().unwrap().clone())
        }
    }

    fn get_quote_size(&mut self) -> Result<u32, SGXError> {
        if self.quote_size.is_none() {
            let size: u32 = 0;
            trace!("ioctl(SGX_IOCTL_GET_DCAP_QUOTE_SIZE): Get DCAP Quote size");
            let ret = unsafe { libc::ioctl(self.fd()?, IOCTL_GET_DCAP_QUOTE_SIZE, &size) };

            if ret < 0 {
                return Err(SGXError::IoctlClientError {
                    request_type: "IOCTL_GET_DCAP_QUOTE_SIZE",
                    ret,
                });
            }

            self.quote_size = Some(size);
        }

        Ok(self.quote_size.as_ref().unwrap().clone())
    }

    pub fn generate_quote(&mut self, report_data: SGXReportData) -> Result<Vec<u8>, SGXError> {
        let mut quote_size = self.get_quote_size()?;
        let mut quote_buf: Vec<u8> = vec![0; quote_size as usize];

        let quote_arg: IoctlGenDCAPQuoteArg = IoctlGenDCAPQuoteArg {
            report_data: &report_data,
            quote_size: &mut quote_size,
            quote_buf: quote_buf.as_mut_ptr(),
        };

        trace!("ioctl(IOCTL_GEN_DCAP_QUOTE): Generate SGX DCAP Quote");
        let ret = unsafe { libc::ioctl(self.fd()?, IOCTL_GEN_DCAP_QUOTE, &quote_arg) };
        if ret < 0 {
            return Err(SGXError::IoctlClientError {
                request_type: "IOCTL_GEN_DCAP_QUOTE",
                ret,
            });
        }
        Ok(quote_buf)
    }

    fn get_supplemental_size(&mut self) -> Result<u32, SGXError> {
        if self.supplemental_size.is_none() {
            let size: u32 = 0;
            trace!("ioctl(IOCTL_GET_DCAP_SUPPLEMENTAL_SIZE): Get Supplemental size");
            let ret = unsafe { libc::ioctl(self.fd()?, IOCTL_GET_DCAP_SUPPLEMENTAL_SIZE, &size) };

            if ret < 0 {
                return Err(SGXError::IoctlClientError {
                    request_type: "IOCTL_GET_DCAP_SUPPLEMENTAL_SIZE",
                    ret,
                });
            }

            self.supplemental_size = Some(size);
        }

        Ok(self.supplemental_size.as_ref().unwrap().clone())
    }

    pub fn verify_quote(&mut self, quote_buf: &[u8]) -> Result<SGXQuoteVerifyResult, SGXError> {
        let supplemental_data_size = self.get_supplemental_size()?;
        let mut result = SGXQuoteVerifyResult::Unspecified;
        let mut status = 1;
        let mut suppl_buf: Vec<u8> = vec![0; supplemental_data_size as usize];

        let verify_arg = IoctlVerDCAPQuoteArg {
            quote_buf: quote_buf.as_ptr(),
            quote_size: quote_buf.len() as u32,
            collateral_expiration_status: &mut status,
            quote_verification_result: &mut result,
            supplemental_data_size,
            supplemental_data: suppl_buf.as_mut_ptr(),
        };

        trace!("ioctl(IOCTL_VER_DCAP_QUOTE): Verify SGX DCAP Quote");
        let ret = unsafe { libc::ioctl(self.fd()?, IOCTL_VER_DCAP_QUOTE, &verify_arg) };
        if ret < 0 {
            return Err(SGXError::IoctlClientError {
                request_type: "IOCTL_VER_DCAP_QUOTE",
                ret,
            });
        }
        Ok(result)
    }
}

impl Drop for IoctlClient {
    fn drop(&mut self) {
        unsafe {
            if let Some(fd) = self.fd {
                libc::close(fd);
            }
        }
    }
}
