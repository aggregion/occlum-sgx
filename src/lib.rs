use std::fmt::Debug;
use std::mem::size_of;
use std::ops::Deref;

#[macro_use]
extern crate lazy_static;

pub use error::SGXError;
use ioctl::IOCTL_CLIENT;
use log::warn;
pub use types::*;

mod constants;
mod error;
mod ioctl;
mod types;

pub struct SGXQuote {
    buf: Vec<u8>,
    report_body: *const SGXReportBody,
}

impl TryFrom<Vec<u8>> for SGXQuote {
    type Error = SGXError;
    fn try_from(buf: Vec<u8>) -> Result<Self, Self::Error> {
        let report_body_offset = size_of::<SGXQuoteHeader>();
        let report_body_size = size_of::<SGXReportBody>();

        if buf.len() < report_body_offset + report_body_size {
            return Err(SGXError::BadQuoteLength {
                min: buf.len(),
                actual: report_body_offset + report_body_size,
            });
        }

        let report_body = buf.as_slice()[report_body_offset..].as_ptr() as *const SGXReportBody;

        Ok(Self { buf, report_body })
    }
}

impl TryFrom<&[u8]> for SGXQuote {
    type Error = SGXError;

    fn try_from(buf: &[u8]) -> Result<Self, Self::Error> {
        buf.to_vec().try_into()
    }
}

impl TryFrom<ReportData> for SGXQuote {
    type Error = SGXError;

    fn try_from(value: ReportData) -> Result<Self, Self::Error> {
        Self::from_report_data(&value)
    }
}

impl Deref for SGXQuote {
    type Target = [u8];

    fn deref(&self) -> &Self::Target {
        self.buf.as_ref()
    }
}

impl SGXQuote {
    // Create a new SGXQuote from ReportData, it needs to be run on the SGX server in an Occlum instance (requires PCCS).
    pub fn from_report_data(data: &ReportData) -> Result<Self, SGXError> {
        let result = IOCTL_CLIENT
            .lock()
            .unwrap()
            .generate_quote(SGXReportData::new(data.clone()))?;
        Ok(result.try_into()?)
    }

    // Restore SGXQuote from slice of bytes.
    pub fn from_slice(slice: &[u8]) -> Result<Self, SGXError> {
        slice.try_into()
    }

    pub fn as_slice(&self) -> &[u8] {
        &*self
    }

    // Verify SGXQuote and return status
    pub fn verify_result(&self) -> Result<SGXQuoteVerifyResult, SGXError> {
        IOCTL_CLIENT.lock().unwrap().verify_quote(self.buf.as_ref())
    }

    // Verify SGXQuote, if it is not valid, return error
    pub fn verify(&self) -> Result<(), SGXError> {
        let result = self.verify_result()?;

        match result {
            SGXQuoteVerifyResult::Ok => Ok(()),
            SGXQuoteVerifyResult::ConfigNeeded
            | SGXQuoteVerifyResult::OutOfDate
            | SGXQuoteVerifyResult::OutOfDateConfigNeeded
            | SGXQuoteVerifyResult::SwHardeningNeeded
            | SGXQuoteVerifyResult::ConfigAndSwHardeningNeeded => {
                warn!(
                    "SGX Quote Verification completed with non-terminal result: {:?}",
                    result
                );
                Ok(())
            }
            _ => Err(SGXError::VerifyQuoteFailed(result)),
        }
    }

    pub fn isv_family_id(&self) -> SGXFamilyId {
        unsafe { (*self.report_body).isv_family_id }
    }

    pub fn isv_ext_prod_id(&self) -> SGXExtProdId {
        unsafe { (*self.report_body).isv_ext_prod_id }
    }

    pub fn config_id(&self) -> SGXConfigId {
        unsafe { (*self.report_body).config_id }
    }

    pub fn mrenclave(&self) -> SGXMeasurement {
        unsafe { (*self.report_body).mr_enclave }
    }

    pub fn mrsigner(&self) -> SGXMeasurement {
        unsafe { (*self.report_body).mr_signer }
    }

    pub fn product_id(&self) -> u16 {
        unsafe { (*self.report_body).isv_prod_id }
    }

    pub fn version(&self) -> u16 {
        unsafe { (*self.report_body).isv_svn }
    }

    pub fn report_data(&self) -> SGXReportData {
        unsafe { (*self.report_body).report_data }
    }
}

impl Debug for SGXQuote {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("SGXQuote")
            .field("mrenclave", &self.mrenclave())
            .field("mrsigner", &self.mrsigner())
            .field("report_body", &self.report_data())
            .field("product_id", &self.product_id())
            .field("version", &self.version())
            .field("family_id", &self.isv_family_id())
            .field("ext_prod_id", &self.isv_ext_prod_id())
            .field("config_id", &self.config_id())
            .finish()
    }
}
