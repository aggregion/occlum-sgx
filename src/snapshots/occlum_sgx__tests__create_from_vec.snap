---
source: src/lib.rs
expression: "format!(\"{:?}\", quote)"
---
"SGXQuote { mrenclave: \"9c90fd81f6e9fe64b46b14f0623523a52d6a5678482988c408f6adffe6301e2c\", mrsigner: \"6d5ead54bfbe9494e1cd9042bb7c25d74c597d4700e332b1b3168a60712c1e02\", report_body: \"AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA\", product_id: 4000, version: 5000, family_id: FamilyId { low: \"0x2534\", high: \"0x3423\" }, ext_prod_id: ExtProdId { low: \"0x423452\", high: \"0x35432\" }, config_id: \"VaHUbdxENJhdwquCHHYG+ytUrqVPeVkj7z/qDjo30lRDmrH9riiZ+1UYmk8cPPZzjn295KVejyt19iBWH7VDow\" }"

